File: frontend/lib/src/model/mood.dart
// Sy (Share your mood with anyone)
// Copyright (C) July 2021 Ivan Usalko <ivict@rambler.ru>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import 'package:frontend/src/model/geometry_shape.dart';

import 'geometry.dart';

/// Basic class
class Mood {
  String id;
  DateTime? created;

  GeometryShape kind;
  List<Geometry?> content;

  Mood(this.id, this.kind, this.content, {this.created});

  factory Mood.fromJson(Map<String, dynamic> json) {
    return Mood(
      json['id'],
      GeometryShapeExt.fromRaw(json['kind'] as String),
      (json['content'] as List<dynamic>)
          .map((e) => e == null ? null : Geometry.fromJson(e))
          .toList(),
      created: json['created'] != null ? DateTime.parse(json['created']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'created': created?.toIso8601String(),
      'kind': GeometryShapeExt.toJson(kind),
      'content': content.map((e) => e != null ? e.toJson() : null).toList()
    };
  }

  static Mood triangle(List<Geometry?> content, {DateTime? created}) {
    return Mood('', GeometryShape.Triangle, content, created: created);
  }

  static Mood square(List<Geometry?> content, {DateTime? created}) {
    return Mood('', GeometryShape.Square, content, created: created);
  }

  static Mood circle(List<Geometry?> content, {DateTime? created}) {
    return Mood('', GeometryShape.Circle, content, created: created);
  }
}

File: backend-java-spring-maven/src/main/java/io/github/usalko/sy/model/Mood.java
// Sy (Share your mood with anyone)
// Copyright (C) July 2021 Ivan Usalko <ivict@rambler.ru>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
package io.github.usalko.sy.model;

import com.fasterxml.jackson.annotation.JsonFormat;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.time.LocalDateTime;
import java.util.List;

@MappedSuperclass
public abstract class Mood<T extends MoodGeometryShape> {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
    private LocalDateTime created;

    @NotNull(message = "Parent geometry form is required.")
    @Basic(optional = false)
    @ManyToOne
    @JoinColumn(name = "geometry_shape_id")
    private GeometryShape geometryShape;

    public LocalDateTime getCreated() {
        return created;
    }

    public void setCreated(LocalDateTime created) {
        this.created = created;
    }

    public GeometryShape getGeometryShape() {
        return geometryShape;
    }

    public void setGeometryShape(GeometryShape geometryShape) {
        this.geometryShape = geometryShape;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public abstract List<T> getMoodGeometryShapes();

    public abstract void setMoodGeometryShapes(List<T> result);
}
